modepy/quadrature/finding.py

```diff
@@ -0,0 +1,423 @@
+        a degree-1 monomial over :math:`\Omega`.
+    :arg onb: An orthonormal basis of functions. Each function takes
+        in an array of (complex) nodes and returns an array
+        of the same shape containing the function values.
+        Functions are expected to be orthogonal with respect to
+        the complex-valued inner product.
+
+        Integrands accepting real-valued node coordinates in two dimensions
+        (shaped ``(2, nnodes)`` may be converted to functions acceptable by
+        this function via :func:`adapt_2d_integrands_to_complex_arg`.)
+    :returns: an array of shape ``(len(onb),)`` containing nodes, complex-valued.
+
+    .. note::
+
+        This function is based on complex arithmetic and thus only
+        usable for domains $\Omega$ with one and two (real-valued)
+        dimensions.
+
+    .. note::
+
+        (Empirically) it seems acceptable for the functions
+        in *onb* to be exclusively real-valued, i.e., particularly,
+        no assumptions on complex differentiability are needed.
+
+    .. note::
+
+        As noted in [Vioreanu2011]_, the returned nodes should obey
+        the symmetry of the domain.
+
+    """
+    n = len(onb)
+    mat = np.array([
+        [
+            integrate(
+                _ProductIntegrand((
+                    _identity_integrand,
+                    onb[i],
+                    _ConjugateIntegrand(onb[j]))))
+            for j in range(n)]
+        for i in range(n)])
+
+    return la.eigvals(mat)
+
+
+def find_weights_undetermined_coefficients(
+            integrands: Sequence[Integrand],
+            nodes: np.ndarray,
+            reference_integrals: np.ndarray,
+        ) -> np.ndarray:
+    """
+    Uses the method of undetermined coefficients to find weights
+    for a quadrature rule using *nodes*, for the provided
+    *integrands*.
+
+    :arg integrands: A sequence of functions that accept an array of nodes
+        of shape either ``(ndim, nnodes)`` or ``(nnodes,)`` and return
+        an array of shape ``(nnodes,)`` with the value of the
+        basis function at the node.
+    :arg nodes: shaped ``(ndim, nnodes)`` or ``(nnodes,)``, real-valued.
+        Must be compatible with *integrands*.
+    :arg reference_integrals: shaped ``(len(integrands),)``
+
+    .. note::
+
+        Tolerates overdetermined systems, will provide the least squares solution.
+
+    .. note::
+
+        Unlike :func:`guess_nodes_vr`, domains of any dimensionality are allowed.
+    """
+
+    if len(reference_integrals) != len(integrands):
+        raise ValueError(
+                "number of integrands must match number of reference integrals")
+    if len(reference_integrals) < len(nodes):
+        from warnings import warn
+        warn("Underdetermined quadrature system", stacklevel=2)
+
+    from modepy import vandermonde
+    return la.lstsq(vandermonde(integrands, nodes).T, reference_integrals)[0]
```

Comment:
Does this fall back to a normal solve if the system is square?

Reply:
Numerically, it's equivalent, in the square, full-rank case. Cost-wise, you're still paying for the SVD.

Reply:
(I don't care much about the square use case here: If someone is using this, they're likely aiming higher than Newton-Cotes.)